lib/ace/bidihandler.js

```diff
@@ -0,0 +1,338 @@
+/* ***** BEGIN LICENSE BLOCK *****
+ * Distributed under the BSD license:
+ *
+ * Copyright (c) 2010, Ajax.org B.V.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Ajax.org B.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL AJAX.ORG B.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+define(function(require, exports, module) {
+"use strict";
+
+var bidiUtil = require("./lib/bidiutil");
+var lang = require("./lib/lang");
+var useragent = require("./lib/useragent");
+
+/**
+ * This object is used to ensure Bi-Directional support (for languages with text flowing from right to left, like Arabic or Hebrew)
+ * including correct caret positioning, text selection mouse and keyboard arrows functioning
+ * @class BidiHandler
+ **/
+
+/**
+ * Creates a new `BidiHandler` object
+ * @param {EditSession} session The session to use
+ *
+ * @constructor
+ **/
+var BidiHandler = function(session) {
+    this.session = session;
+    this.bidiMap = {};
+    /* current screen row */
+    this.currentRow = null;
+    this.bidiUtil = bidiUtil;
+    /* Arabic/Hebrew character width differs from regular character width */
+    this.charWidths = [];
+    this.EOL;
```

Comment:
- bidiUtil references lightweight Bidi engine used to produce Logical <-> Visual maps. These maps are used to correctly calculate caret position.
- charWidths contains widths of various characters (English, Bidi, Zero width, Arabic Lam-Alef). Again, they are used to correctly calculate caret position.
- EOL - end of line character, it may vary in ACE. Line may be followed by end-of-line or end-of-file symbol (when show invisibles) and they are used in caret position calculations (sometimes EOL occurs inside the Bidi line, not at the end)